### PR TITLE
Add MCAD API groups migration plan

### PR DESCRIPTION
**Pull request by astefanutti**
Closes project-codeflare/multi-cluster-app-dispatcher#352.

### CODE REVIEW

File: PCF-ADR-0005-mcad-api-groups-migration.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+Some of the MCAD APIs, mainly the AppWrapper API, is used by downstream projects, as well as end-users.
+From a client standpoint, the impacts of an API group change imply updating the `apiVersion` field of all the MCAD resources they define, as well as updating the group segment of the REST endpoints, created from the CRDs, that they consume.
+
+A migration plan is needed to provide these clients with a smooth transition to the new API groups.
+
+## Goals
+
+The goal is to inform a _progressive_ migration path, i.e., one that does not require all the downstream projects to handle the impacts in locked steps with MCAD.
+A project like [TorchX](https://github.com/pytorch/torchx), which integrates with MCAD[^1] via the AppWrapper API, has its own release cadence, and it’d be very challenging to have all the components migrated within a single migration cycle, without breaking compatibility.
```
Comment (asm582) [inline PCF-ADR-0005-mcad-api-groups-migration.md]: I think we should also include KubeRay as downstream component consuming MCAD: https://ray-project.github.io/kuberay/guidance/kuberay-with-MCAD/

File: PCF-ADR-0005-mcad-api-groups-migration.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+The new API groups are introduced in MCAD, along with the existing ones.
+Functionally, the MCAD and InstaScale controllers are updated, so they can equally reconcile the old and new APIs.
+
+To implement that dual API groups support, it may be necessary to duplicate the MCAD API packages, as well as the generated client packages.
+Indeed, [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) APIs, which are used by InstaScale, are designed to use the API structs, and require the mapping between the API structs and the Group/Version/Kind(GVK)s to be an injection.
+Otherwise stated, controller-runtime mandates the scheme holding the mapping between the API structs and the GVKs to have a single GVK entry per API struct.
+
+This boils down to the following tasks:
+
+* Duplicate the API packages, with the new API groups, and mark the old ones as deprecated
```
Comment (asm582) [inline PCF-ADR-0005-mcad-api-groups-migration.md]: With the desire to make migration easier, is there a possibility to declare a current target MCAD release compatible with "all" downstream projects for the short term and not do duplication? 
Most downstream integrations are managed by the MCAD team, later individual integrators can work on the migration path.

File: PCF-ADR-0005-mcad-api-groups-migration.md (hunk 1 above)
Comment (astefanutti) [reply PCF-ADR-0005-mcad-api-groups-migration.md]: Done with f7126ca7f30d2e51a090be9b745622a05481836d. It's not a technical impact, nor there are any references to the existing API groups. Thought we can take that opportunity to update the links to the CodeFlare project.

File: PCF-ADR-0005-mcad-api-groups-migration.md (hunk 2 above)
Comment (astefanutti) [reply PCF-ADR-0005-mcad-api-groups-migration.md]: I agree we should lower duplication as much as possible.

I did not realize the TorchX MCAD scheduler is currently delivered from the CodeFlare fork of TorchX. That makes all the impacted components under our control, as you said.

I've added the _one-shot_ migration alternative in 5735b236e46708af4591aaa08c57a5b19eddbbcb, and given and example of how that could go.

File: PCF-ADR-0005-mcad-api-groups-migration.md (hunk 1 above)
Comment (astefanutti) [reply PCF-ADR-0005-mcad-api-groups-migration.md]: If the consensus is to go with the _on-shot_ migration alternative, I'll flip it with the _progressive_ migration proposal.

File: PCF-ADR-0005-mcad-api-groups-migration.md (hunk 2 above)
Comment (astefanutti) [reply PCF-ADR-0005-mcad-api-groups-migration.md]: If the consensus is to go with the _one-shot_ migration alternative, I'll flip it with the _progressive_ migration proposal.

Comment (asm582) [reply PCF-ADR-0005-mcad-api-groups-migration.md]: If one-shot migration is simpler, I think we should bite the bullet and I vote for doing it. a stop-gap solution is to ask users to use let's say release 1.32 which has old CRDs.

File: PCF-ADR-0005-mcad-api-groups-migration.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
```
Comment (z103cb) [inline PCF-ADR-0005-mcad-api-groups-migration.md]: I think that would be good for this document that the package names / structures are presented in this ADR.

Also, having support in MCAD for two api groups might be a lot of work and make the existing codebase even more incomprehensible. Is it worth adding a "conversion" web hook so that the changes to MCAD codebase would be more 'sane' ? 

If yes, does it make sense to explicitly mention it here ?

File: PCF-ADR-0005-mcad-api-groups-migration.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+The KubeRay documentation can be updated independently.
+
+## Stakeholder Impacts
+
+| Group                         | Key Contacts       | Date       | Impacted? |
+| ----------------------------- | ------------------ | ---------- | --------- |
+| CodeFlare SDK                 | Mustafa Eyceoz     |            | yes       |
+| MCAD                          | Abhishek Malvankar |            | yes       |
+| InstaScale                    | Abhishek Malvankar |            | yes       |
+| CodeFlare Operator            | Anish Asthana      |            | no        |
```
Comment (z103cb) [inline PCF-ADR-0005-mcad-api-groups-migration.md]: Can you add me for the KubeRay apiserver and KubeRay documentation ? If the implementation of this change will trail, there will be an impact on the kuberayapi server. 

TL;DR. Kueberay api server will submit appwrappers for the KubeRay CRDS. We plan on depending on the MCAD api package to programately create the appwrappers.

**Review by z103cb — COMMENTED** [review]
Overall LGTM. Left some questions

File: PCF-ADR-0005-mcad-api-groups-migration.md (hunk 3 above)
Comment (astefanutti) [reply PCF-ADR-0005-mcad-api-groups-migration.md]: Conversion webhooks address the conversion across versions of the same CRD. The `.spec.group` field is immutable, as the other fields that uniquely identify the CRD, e.g., the name, which implies the change of API group is actually the creation of an entirely new CRD, completely unrelated to the existing CRD. So conversion webhooks are not applicable in that case.

File: PCF-ADR-0005-mcad-api-groups-migration.md (hunk 4 above)
Comment (astefanutti) [reply PCF-ADR-0005-mcad-api-groups-migration.md]: Added with 5473484368460890188cba3e732ef9ffcaa9606f.

**Comment by tardieu** [comment]
If we are going to force every downstream project/user to take action, it may be worth considering making other/all desirable changes to the CRD structure at the same time, so that we don't have to repeat the disruption again and again. Maybe we need first separate discussion/ADR about these changes and then decide whether to combine these efforts.

**Comment by astefanutti** [comment]
> If we are going to force every downstream project/user to take action, it may be worth considering making other/all desirable changes to the CRD structure at the same time, so that we don't have to repeat the disruption again and again. Maybe we need first separate discussion/ADR about these changes and then decide whether to combine these efforts.

@tardieu that's a good point. It's worth noting that those "structural" changes, within the CRDs, could be handled with the usual CRD versioning mechanism, possibly using conversion webhooks (or CEL based conversion in the future) to automate it, as opposed to changes to identifying fields, like the API group.

That being said, I agree it'd be worth assessing if any major breaking changes could be piggy-backed with that effort.

@asm582 @z103cb could you please provide your input here?

**Comment by asm582** [comment]
> > If we are going to force every downstream project/user to take action, it may be worth considering making other/all desirable changes to the CRD structure at the same time, so that we don't have to repeat the disruption again and again. Maybe we need first separate discussion/ADR about these changes and then decide whether to combine these efforts.
> 
> @tardieu that's a good point. It's worth noting that those "structural" changes, within the CRDs, could be handled with the usual CRD versioning mechanism, possibly using conversion webhooks (or CEL based conversion in the future) to automate it, as opposed to changes to identifying fields, like the API group.
> 
> That being said, I agree it'd be worth assessing if any major breaking changes could be piggy-backed with that effort.
> 
> @asm582 @z103cb could you please provide your input here?

Thanks, @astefanutti  and @tardieu - "structural changes" as mentioned are bound to happen in the future. For the features that are required for the immediate release, I don't think there is anything major that needs to be added to the effort.

**Comment by z103cb** [comment]
> > If we are going to force every downstream project/user to take action, it may be worth considering making other/all desirable changes to the CRD structure at the same time, so that we don't have to repeat the disruption again and again. Maybe we need first separate discussion/ADR about these changes and then decide whether to combine these efforts.
> 
> @tardieu that's a good point. It's worth noting that those "structural" changes, within the CRDs, could be handled with the usual CRD versioning mechanism, possibly using conversion webhooks (or CEL based conversion in the future) to automate it, as opposed to changes to identifying fields, like the API group.
> 
> That being said, I agree it'd be worth assessing if any major breaking changes could be piggy-backed with that effort.
> 
> @asm582 @z103cb could you please provide your input here?


I think is OK to try to piggy back everything if all the items are known at this point in time. From my perspective (KubeRay API server integration) the sooner the final API is available the better. 

The only items that I can think that would fall into `structural` category are:
* [Is enabling status sub resource for the Quota Management CRD(s)](https://github.com/project-codeflare/multi-cluster-app-dispatcher/issues/389) (even if is just empty and not populated in the first couple of releases). 
* [Move AppWrapper quota request from label to element of AppWrapper object](https://github.com/project-codeflare/multi-cluster-app-dispatcher/issues/377)

File: PCF-ADR-0005-mcad-api-groups-migration.md (hunk 2 above)
Comment (anishasthana) [reply PCF-ADR-0005-mcad-api-groups-migration.md]: I agree, I'm a +1 for one-shot

Comment (z103cb) [reply PCF-ADR-0005-mcad-api-groups-migration.md]: +1 on that also.

Comment (astefanutti) [reply PCF-ADR-0005-mcad-api-groups-migration.md]: Thanks all for the feedback, I've promoted the _one-shot_ migration as the proposed plan in 2c3eefe2cf8c7bff02923a7bfcba204db4e8bdbd.

**Comment by Sara-KS** [comment]
For the TorchX+MCAD piece, I plan to try to support both the old version and the new version to help with the transition for users on different systems. Is there a clear way to tell which CRDs are in use that does not require administrative type permissions?

**Comment by astefanutti** [comment]
> For the TorchX+MCAD piece, I plan to try to support both the old version and the new version to help with the transition for users on different systems. Is there a clear way to tell which CRDs are in use that does not require administrative type permissions?

One possible solution would be to query the discovery API, and check if the new API group is present. Generally most users are granted permission to query the discovery API, which is what's used when running the `kubectl api-resources` command.

**Review by z103cb — APPROVED** [review]
LGTM, but I would wait for the others to chime in before considering the ADR approved. @tardieu needs to sign off on this ADR.

**Comment by tardieu** [comment]
+1 for one-shot.

**Review by anishasthana — APPROVED** [review]
Approved from my perspective.

**Comment by z103cb** [comment]
Approved.

**Review by dimakis — APPROVED** [review]
LGTM, makes logical sense.

**Comment by anishasthana** [comment]
I think we have quorum for the approvals. The work has already somewhat begun -- I'm going to go ahead and merge this. If in future we require additional revisions, we can always propose a new ADR / add an addendum to this one.
